### PR TITLE
Fix issue #53: Illegal invocation in Jest tests

### PR DESCRIPTION
**Pull request by ilanashapiro**
This PR fixes https://github.com/tree-sitter/node-tree-sitter/issues/53

This PR is a modification and update to @cellog's PR from a few years ago: https://github.com/tree-sitter/node-tree-sitter/pull/75.

From @cellog:
"Basically, the problem described in https://stackoverflow.com/questions/9677985/uncaught-typeerror-illegal-invocation-in-chrome is what is happening here, but it appears to be a race condition.

When we re-load the context in the same process (which happens with Jest's worker pool), the getter is triggered, for no apparent reason, BEFORE the native extension has had a chance to load. this is actually the user-land object, with no knowledge of the native prototype yet. Just by changing the name of the extension's getters, it fixes this race condition."

-----

When I tried changing the name of the extension's getters, it resulted in all of them being undefined. The solution I found is to only call the prototype's methods if the native extension is fully loaded. When the extension has fully loaded, "this" should be an instance of the class whose prototype we are acting on. For instance, in the following code we are acting on the prototype of Tree:
```
Object.defineProperty(Tree.prototype, 'rootNode', {
  get() {
    if (this instanceof Tree && rootNode) {
      return unmarshalNode(rootNode.call(this), this);
    }
  },
  configurable: true
});
```
so "this" should be an instance of Tree. If it is not, the native extension has not fully loaded, so we skip the invocation (hence the if statement). Furthermore, I found that even when the context is correct, the race condition sometimes results in the method in question (here, rootNode) being undefined, so the null method check was also added. Finally, I am adding  configurable: true since I found that sometimes, Jest attempts to overwrite properties (specifically, nodeText, but possibly others) on the previous context due to the race condition, and we do not want to throw an error if this happens.

This all results in get() returning `undefined` when the native extension isn't fully loaded, so the jest tests need to have a check to reflect this, for instance:
```
it("should work", () => {
    const code = jsParser.parse(constants.INPUT);
    if (code.rootNode) { // native extension has finished loading
      const output = code.rootNode.toString();
      expect(output).toBe(constants.OUTPUT);
    }
  });
```

There are 3 Jest tests currently in the repo, you can ensure this PR works by simply running `npm test`.

### CODE REVIEW

**Comment by ilanashapiro** [comment]
@rattrayalex @RobertCraigie

**Comment by ilanashapiro** [comment]
The tests on Node 10/12 are failing due to the version of Jest (the latest) I have in package.json. I will try to figure out which version of Jest is compatible with Node 10 and up.

**Comment by ilanashapiro** [comment]
@verhovsky Could I ask you or a member of your team to take a look at this? Thank you!

**Comment by ilanashapiro** [comment]
@verhovsky Just wanted to check in if it would be possible to get a review on this? Thank you!

File: index.js [hunk 1]
```diff
@@ -20,20 +20,37 @@ const {rootNode, edit} = Tree.prototype;
 
 Object.defineProperty(Tree.prototype, 'rootNode', {
   get() {
-    return unmarshalNode(rootNode.call(this), this);
-  }
+    /*
+      Due to the race condition arising from Jest's worker pool, "this"
```
Comment (verhovsky) [inline index.js:24]:
```suggestion
      Due to a race condition arising from Jest's worker pool, "this"
```

File: index.js [hunk 2]
```diff
@@ -20,20 +20,37 @@ const {rootNode, edit} = Tree.prototype;
...
+      has no knowledge of the native extension if the extension has not
+      yet loaded when multiple Jest tests are being run simultaneously.
+      If the extension has correctly loaded "this" should be an instance
```
Comment (verhovsky) [inline index.js:27]:
```suggestion
      If the extension has correctly loaded, "this" should be an instance 
```

File: index.js [hunk 3]
```diff
@@ -20,20 +20,37 @@ const {rootNode, edit} = Tree.prototype;
...
-  }
+    /*
+      Due to the race condition arising from Jest's worker pool, "this"
+      has no knowledge of the native extension if the extension has not
+      yet loaded when multiple Jest tests are being run simultaneously.
+      If the extension has correctly loaded "this" should be an instance 
+      of the class whose prototype we are acting on (in this case, Tree).
+      Furthermore, the race condition sometimes results in the function in 
+      question being undefined even when the context is correct, so we also 
+      add the null function check.
```
Comment (verhovsky) [inline index.js:31]:
```suggestion
      perform a null function check.
```

**Pull request merged**